File: LeetCode_ProblemSet/Go/Problem1508.go
package main

import "sort"

func rangeSum(nums []int, n int, left int, right int) int {
	prefixSums := getPrefixSum(nums, n)
	subarraySums := getSubarraySums(prefixSums, n)
	return getSortedSumOverRange(subarraySums, left, right)
}

func getPrefixSum(nums []int, n int) []int {
	prefixSums := make([]int, n+1)
	prefixSum := 0
	for i := range n {
		prefixSum += nums[i]
		prefixSums[i+1] = prefixSum
	}
	return prefixSums
}

func getSubarraySums(prefixSums []int, n int) []int {
	subarraySums := make([]int, (n*(n+1))/2)
	idx := 0
	for i := range n {
		for j := i; j < n; j++ {
			subarraySums[idx] = prefixSums[j+1] - prefixSums[i]
			idx++
		}
	}
	return subarraySums
}

func getSortedSumOverRange(subarraySums []int, left, right int) int {
	sort.Slice(subarraySums, func(i, j int) bool {
		return subarraySums[i] < subarraySums[j]
	})
	sum := 0
	for i := left -1; i < right; i++ {
		sum = (sum + subarraySums[i]) % (1e9 + 7)
	}
	return sum
}

File: LeetCode_ProblemSet/Problem1508.py
class Solution(object):
    def rangeSum(self, nums, n, left, right):
        """
        :type nums: List[int]
        :type n: int
        :type left: int
        :type right: int
        :rtype: int
        """
        prefix_sums = [0] * (n + 1)
        prefix_sum = 0
        
        for i in range(n):
            prefix_sum += nums[i]
            prefix_sums[i + 1] = prefix_sum
        
        
        subarray_sums = []
    
        for i in range(n):
            for j in range(i, n):
                subarray_sums += [prefix_sums[j + 1] - prefix_sums[i]]
                
        subarray_sums = sorted(subarray_sums)
        ans = 0
        
        for i in range(left - 1, right):
            ans = (ans + subarray_sums[i]) % (1e9 + 7)
        
        return ans
        